### PR TITLE
NIFI-7981 - add support for enum type in avro schema

### PR DESCRIPTION
**Pull request by pvillard31**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [x] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java [hunk 1]
```diff
@@ -1025,6 +1029,17 @@ public static boolean isStringTypeCompatible(final Object value) {
         return value != null;
     }
 
+    public static boolean isEnumTypeCompatible(final Object value, final EnumDataType enumType) {
+        return enumType.getEnums() != null && enumType.getEnums().contains(value);
+    }
+
+    private static Object toEnum(Object value, EnumDataType dataType, String fieldName) {
+        if(dataType.getEnums() != null && dataType.getEnums().contains(value)) {
```
Comment (MikeThomsen) [inline nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java:1037]: Are you assuming only conversion from strings here? Like not even allowing people to convert booleans to TRUE and FALSE?

File: nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/json/JsonTreeRowRecordReader.java [hunk 2]
```diff
@@ -84,22 +84,6 @@ private JsonNode getChildNode(final JsonNode jsonNode, final RecordField field)
         return null;
     }
 
-    private boolean isDateTimeTimestampType(final RecordField field) {
```
Comment (MikeThomsen) [inline nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/json/JsonTreeRowRecordReader.java:87]: Was this just unused code?

**Review by MikeThomsen — COMMENTED** [review]
Overall LGTM. Will kick the tires and see how it performs.

File: nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/json/JsonTreeRowRecordReader.java (hunk 2 above)
Comment (pvillard31) [reply nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/json/JsonTreeRowRecordReader.java:87]: yes

File: nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java (hunk 1 above)
Comment (pvillard31) [reply nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java:1037]: I assumed that we can only have strings: http://avro.apache.org/docs/current/spec.html#Enums

Comment (MikeThomsen) [reply nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java:1037]: Yeah, I know the spec only allows strings, but I was wondering if wanted to allow boolean values to be converted to strings here.

**Review by MikeThomsen — APPROVED** [review]

**Comment by MikeThomsen** [comment]
Ran it with a live flow, and it worked as expected with my enum test schema (with both good and bad json inputs).

File: nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java (hunk 1 above)
Comment (markap14) [reply nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java:1037]: We should be very careful here - there is no spec. We absolutely do not - and should not - follow the Avro spec. We should do what makes sense for us. Personally I am not a fan of introducing Enums at all. I very intentionally didn't add those in as a data type when we created the initial set.  More on that in a general comment below :)

**Comment by markap14** [comment]
I'm personally not a fan of introducing Enum here as a datatype. I avoided introducing that as one of the data types when originally creating the Record API. What I think makes far more sense is a more general notion of restrictions of types. Consider, for example, how XSD allows you to add a restriction to a base type. We want to eventually support specifying schemas using XSD, and we'll need to support this notion in order to make that happen. I would certainly favor introducing that notion now over introducing an Enum. Then, when converting an Avro schema into a RecordSchema, we could do so as a restriction on the String type. This provides far more power and flexibility when using schemas, such as specifying that an int must be in the range of 1-100, and this then becomes extremely powerful when coupled with something like ValidateRecord.
Adding an enum at this time, IMO, makes this a bit awkward, because it is then a specific type whose purpose is to introduce a restriction on another type (I.e., it is basically a Restriction on a String type). So I think it kind of makes things confusing if our API provides two mechanisms for specifying an enumeration: using an enum type, or specifying a restriction on a String type...

**Comment by MikeThomsen** [comment]
@markap14 has any work, even just planning, started on restricted types? I agree with your points, and if there are plans to proceed on such a feature set it would make sense to back this type out and plan to massage enum functionality into that.

**Comment by markap14** [comment]
So this isn't really how i was envisioning  the "enum" concept being implemented. But after thinking about it for a bit, I do think it's quite reasonable to have an Enum data type. Even if we introduce a notion of adding restrictions. So I think I'm ok with it. Thanks @pvillard31 and @MikeThomsen !

**Pull request closed without merging**